### PR TITLE
DrivAerML: Depth scaling 5L/512d — compound depth + width

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
4L/512d (4.619%) beats 4L/320d (5.027%) — width scales. If depth also helps (as on AirfRANS where 4L beat 3L by 22%), then 5L/512d could push even further. The 180-min budget allows enough epochs even with the slower per-epoch time.

## Instructions
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 5 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --agent shouko \
  --wandb-name "shouko/drivaerml-5L512d" \
  --wandb-group "drivaerml-depth-scaling"
```

Report val_primary/surface_rel_l2_pct, best epoch, total epochs, time per epoch.

## Baseline
Current best: **4.619%** at epoch 256 (4L/512d, 267 epochs, 180-min)
External target: <3.71%

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT shouko:

## Results

### DrivAerML: 5L/512d depth scaling (180-min budget)

| Metric | 5L/512d (this run) | 4L/512d baseline | Change |
|--------|-------------------|-----------------|--------|
| **val_primary/surface_rel_l2_pct** | **11.88%** (best, epoch 52) | **4.619%** (epoch 256) | +157% ❌ |
| test_primary/surface_rel_l2_pct | 41.73% (final, broken) | N/A | N/A |
| Total epochs | 266 | 267 | comparable |
| Time per epoch | 0.68 min | 0.67 min | comparable |
| Peak GPU memory | 7.2 GB | N/A | — |
| External target | <3.71% | <3.71% | — |

**W&B run:** [28j5ooth](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/28j5ooth)

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 5 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent shouko --wandb-name "shouko/drivaerml-5L512d" --wandb-group "drivaerml-depth-scaling"
```

### What happened — NEGATIVE RESULT

**5L/512d without grad clipping is catastrophically unstable on DrivAerML.** The hypothesis that depth scaling would compound with width scaling is **falsified** in the absence of gradient stabilization.

**Phase 1 (epochs 1-52): Promising convergence**
- The model converged faster than 4L/512d in the early phase: 11.88% at epoch 52
- Grad norms were healthy (0.8-1.2), loss steadily decreasing (1.04 → 0.033)
- Each cosine trough delivered ~1% improvement (29.5% → 18.8% → 15.5% → 12.6% → 11.9%)

**Phase 2 (epochs 55-266): Catastrophic gradient instability**
- **23 instability events** over 211 epochs — approximately one every 9 epochs
- First spike at epoch 55: val jumped 12.5% → 60.1% (back to epoch-1 performance)
- Gradient norm divergence was exponential and terminal:
  - Epoch 52: grad_norm = **0.8** (healthy)
  - Epoch 100: grad_norm = **139** (100x)
  - Epoch 150: grad_norm = **52,353** (65,000x)
  - Epoch 200: grad_norm = **5,231,122** (6.5M x)
  - Epoch 266: grad_norm = **389,918,258** (490M x)
- The model never recovered to its pre-instability best (11.88%) after the first spike
- Post-instability, the model oscillated in a 20-40% band, occasionally spiking to 60%+
- The final 100 epochs were spent in a permanently broken state with grad norms in the millions

**Why it failed:** The 5th transformer layer amplifies gradient magnitudes through the deeper backpropagation path. Without gradient clipping, cosine LR restarts push the model out of basins, and the deeper architecture cannot recover gracefully. The 4L/512d baseline likely sits just below the stability threshold at lr=5e-4 with T_max=30. Adding one more layer crosses that threshold.

**Key evidence:** The 4L/512d baseline ran 267 epochs with stable grad norms and achieved 4.619%. This run achieved comparable throughput (266 epochs, 0.68 min/epoch) but gradient instability destroyed all gains after epoch 52.

### Suggested follow-ups

1. **5L/512d + grad clipping (gc=1.0 or gc=1.5)** — This is the highest-priority follow-up. AirfRANS showed that grad clipping is essential for stabilizing deeper architectures. The pre-instability trajectory (11.88% at epoch 52) suggests the architecture has capacity — it just needs gradient stabilization. gc=1.0 would prevent the spikes while preserving the fast early convergence.

2. **5L/512d + lower LR (3e-4)** — A lower learning rate would reduce the magnitude of cosine peak perturbations, potentially avoiding the instability altogether. The tradeoff is slower convergence.

3. **5L/512d + gc + WD=1e-2** — Compounding all three stability mechanisms (clipping + weight decay + depth) could unlock the full depth-scaling benefit that the initial convergence trajectory hinted at.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2803 because ADVISOR: Closing — your 5L/512d result of 11.878% diverged, far from the 4.619% baseline. Going deeper (5L) on DrivAerML causes instability. The 4L/512d architecture with stability controls remains the best path. Thank you, shouko..
